### PR TITLE
Add Chainguard base image detection and SBOM reuse

### PR DESCRIPTION
**Pull request by vpetersson**
## Summary

- Detects Chainguard images automatically — both direct (`cgr.dev/chainguard/...`) and as base images in user-built images via BuildKit SLSA provenance attestations
- Downloads the Chainguard-provided SPDX SBOM from cosign attestations (per-architecture, ~122 packages with `pkg:apk/wolfi/...` PURLs)
- Converts SPDX → CycloneDX when CDX format is requested (small SBOM, straightforward mapping)
- Warns users that binaries added via `COPY --from=...` need to be declared with `ADDITIONAL_PACKAGES`
- Bundles `crane` (v0.21.5) and `cosign` (v3.0.6) in the Docker image for registry inspection and attestation retrieval
- Falls through to normal cdxgen/syft generation when no Chainguard image is detected

## Test plan

- [ ] Unit tests pass: `uv run pytest tests/test_chainguard.py -v` (20 tests, all mocked)
- [ ] Full test suite passes with no regressions
- [ ] E2E: `sbomify-action --docker-image cgr.dev/chainguard/python:latest -o sbom.json` produces valid CycloneDX
- [ ] E2E: `sbomify-action --docker-image cgr.dev/chainguard/python:latest -f spdx -o sbom.spdx.json` uses SPDX directly
- [ ] E2E: `sbomify-action --docker-image sbomifyhub/sbomify:latest -o sbom.json` detects Chainguard base via provenance
- [ ] E2E: `sbomify-action --docker-image nginx:latest -o sbom.json` falls through to normal generation
- [ ] Docker build succeeds with crane + cosign installed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: sbomify_action/_generation/chainguard.py [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+        return None
+
+    # Fetch and parse the provenance blob
+    try:
+        # Need to resolve the registry/repo for blob fetching
+        if "@" in docker_image:
+            registry_repo = docker_image.split("@")[0]
+        else:
+            registry_repo = docker_image.split(":")[0]
+        provenance_json = _run_crane(["blob", f"{registry_repo}@{provenance_digest}"])
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/_generation/chainguard.py:249]: Addressed in 65d27be.

File: sbomify_action/cli/main.py [hunk 2]
```diff
@@ -1127,15 +1129,53 @@ def run_pipeline(config: Config) -> None:
...
+                        with open(STEP_1_FILE, "w", encoding="utf-8") as f:
+                            json.dump(spdx_sbom, f, ensure_ascii=False)
+
+                    result = GenerationResult.success_result(
+                        output_file=STEP_1_FILE,
+                        sbom_format=config.sbom_format,
+                        spec_version=config.spec_version
+                        or ("1.6" if config.sbom_format == "cyclonedx" else "2.3"),
+                        generator_name="chainguard-sbom",
+                    )
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py]: Addressed in 65d27be.

File: sbomify_action/_generation/chainguard.py [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+
+    machine = platform.machine()
+    arch_map = {
+        "x86_64": "amd64",
+        "aarch64": "arm64",
+        "arm64": "arm64",
+    }
+    arch = arch_map.get(machine, machine)
+    return f"linux/{arch}"
+
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/_generation/chainguard.py]: Addressed in 65d27be.

File: sbomify_action/_generation/chainguard.py [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+        return None
+
+    # Fetch the attestation manifest to find the provenance layer
+    try:
+        att_manifest_json = _run_crane(["manifest", f"{docker_image.split(':')[0]}@{att_digest}"])
+        # Handle case where docker_image has @ instead of :
+        if "@" in docker_image:
+            registry_repo = docker_image.split("@")[0]
+            att_manifest_json = _run_crane(["manifest", f"{registry_repo}@{att_digest}"])
+        att_manifest = json.loads(att_manifest_json)
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/_generation/chainguard.py:230]: Addressed in 65d27be.

File: sbomify_action/cli/main.py [hunk 5]
```diff
@@ -1127,15 +1129,56 @@ def run_pipeline(config: Config) -> None:
...
+                        "will NOT be included. Use ADDITIONAL_PACKAGES to declare them. "
+                        "See: https://github.com/sbomify/sbomify-action#additional-packages",
+                        title="Chainguard Image Detected",
+                    )
+
+                    try:
+                        spdx_sbom = fetch_chainguard_sbom(chainguard_info)
+                    except RuntimeError as e:
+                        logger.warning(f"Failed to fetch Chainguard SBOM, falling back to normal generation: {e}")
+                        chainguard_info = None
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py:1160]: Addressed in b9226bc — moved the warning to after `fetch_chainguard_sbom()` succeeds.

File: sbomify_action/_generation/chainguard.py [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+    """
+    if not shutil.which("cosign"):
+        raise RuntimeError("cosign not found on PATH, cannot fetch Chainguard SBOM")
+
+    image_with_digest = f"{info.image_ref}@{info.digest}"
+    logger.info(f"Fetching Chainguard SBOM for {image_with_digest}")
+
+    try:
+        output = _run_cosign(["download", "attestation", image_with_digest])
+    except subprocess.CalledProcessError as e:
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/_generation/chainguard.py:324]: Intentionally using `cosign download attestation` rather than `verify-attestation`. Verification would require pinning Chainguard's OIDC issuer and certificate identity, which are implementation details that may change. Our goal here is SBOM reuse, not supply chain verification — the user can separately verify their images with cosign if needed. Updated README to say "Chainguard-provided" rather than "signed".

File: sbomify_action/_generation/chainguard.py [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+        else:
+            comp_type = ComponentType.LIBRARY
+
+        # Build component
+        comp = Component(
+            type=comp_type,
+            name=name,
+            version=version if version else None,
+            bom_ref=spdx_id,
+        )
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/_generation/chainguard.py:431]: The `Component` constructor accepts `str` for `bom_ref` and wraps it in `BomRef` internally — no downstream mismatch. No other code in the main source passes `BomRef` explicitly either.

File: sbomify_action/cli/main.py [hunk 8]
```diff
@@ -1127,15 +1129,56 @@ def run_pipeline(config: Config) -> None:
...
+                                f.write(cdx_json)
+                        else:
+                            with open(STEP_1_FILE, "w", encoding="utf-8") as f:
+                                json.dump(spdx_sbom, f, ensure_ascii=False)
+
+                        result = GenerationResult.success_result(
+                            output_file=STEP_1_FILE,
+                            sbom_format=config.sbom_format,
+                            spec_version=config.spec_version or ("1.6" if config.sbom_format == "cyclonedx" else "2.3"),
+                            generator_name="chainguard-sbom",
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py]: Good catch — fixed in 5bdf512. Now extracts the actual SPDX version from the document (e.g. `SPDX-2.3` → `2.3`) instead of using `config.spec_version`.

File: sbomify_action/_generation/chainguard.py [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+            continue
+
+        payload_b64 = envelope.get("payload", "")
+        if not payload_b64:
+            continue
+
+        try:
+            payload = json.loads(base64.b64decode(payload_b64))
+        except (json.JSONDecodeError, Exception):
+            continue
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/_generation/chainguard.py:345]: Already addressed in 3342300 — narrowed to `(json.JSONDecodeError, ValueError, UnicodeDecodeError)`.

File: sbomify_action/cli/main.py [hunk 10]
```diff
@@ -1127,15 +1129,56 @@ def run_pipeline(config: Config) -> None:
...
+                    fetch_chainguard_sbom,
                 )
-                if not result.success:
-                    raise SBOMGenerationError(result.error_message or "SBOM generation failed")
+
+                chainguard_info = detect_chainguard_image(config.docker_image)
+                if chainguard_info:
+                    logger.info(f"Detected Chainguard base image: {chainguard_info.image_ref}")
+
+                    try:
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py]: The Chainguard logic itself has 26 unit tests covering detection, fetch, conversion, and error paths. The CLI integration path is thin orchestration (detect → fetch → write file) and is exercised by E2E testing. Adding CLI-level mocked tests for this would duplicate the unit test coverage without meaningful benefit.

File: sbomify_action/cli/main.py [hunk 11]
```diff
@@ -1127,15 +1129,60 @@ def run_pipeline(config: Config) -> None:
...
+                            "Any binaries added via COPY --from=... (e.g. cosign, crane, osv-scanner) "
+                            "will NOT be included. Use ADDITIONAL_PACKAGES to declare them. "
+                            "See: https://github.com/sbomify/sbomify-action#additional-packages",
+                            title="Chainguard Image Detected",
+                        )
+                        if config.sbom_format == "cyclonedx":
+                            cdx_json = convert_spdx_to_cyclonedx(spdx_sbom, config.spec_version or "1.6")
+                            with open(STEP_1_FILE, "w", encoding="utf-8") as f:
+                                f.write(cdx_json)
+                            actual_spec_version = config.spec_version or "1.6"
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py]: Fixed in ba6e23c — CDX < 1.3 now falls back to normal generation with an info log.

File: sbomify_action/_generation/chainguard.py [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+    # Set main component in metadata
+    if main_component:
+        bom.metadata.component = main_component
+
+    # Add all other components
+    for comp in components:
+        bom.components.add(comp)
+
+    # Serialize
+    return serialize_cyclonedx_bom(bom, spec_version)
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/_generation/chainguard.py:475]: Handled at the call site in ba6e23c — the CLI now checks the version before calling `convert_spdx_to_cyclonedx()` and falls back for < 1.3.

File: sbomify_action/cli/main.py [hunk 13]
```diff
@@ -1127,15 +1129,60 @@ def run_pipeline(config: Config) -> None:
...
+                        else:
+                            with open(STEP_1_FILE, "w", encoding="utf-8") as f:
+                                json.dump(spdx_sbom, f, ensure_ascii=False)
+                            # Use the actual SPDX version from the document, not what user requested
+                            spdx_version = str(spdx_sbom.get("spdxVersion", "SPDX-2.3"))
+                            actual_spec_version = spdx_version.replace("SPDX-", "")
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py]: Fixed in ba6e23c — SPDX 3.0.1 requests now skip Chainguard reuse and fall back to normal generation.

File: sbomify_action/cli/main.py [hunk 14]
```diff
@@ -1127,15 +1129,78 @@ def run_pipeline(config: Config) -> None:
...
+                                with open(STEP_1_FILE, "w", encoding="utf-8") as f:
+                                    f.write(cdx_json)
+                                actual_spec_version = spec
+                        else:
+                            with open(STEP_1_FILE, "w", encoding="utf-8") as f:
+                                json.dump(spdx_sbom, f, ensure_ascii=False)
+                            # Use the actual SPDX version from the document
+                            spdx_version = str(spdx_sbom.get("spdxVersion", "SPDX-2.3"))
+                            actual_spec_version = spdx_version.replace("SPDX-", "")
+
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py]: Fixed in 7c467b3 — added a warning log when the Chainguard SPDX version differs from the requested version. SPDX 3.0.1 was already handled (falls back to normal generation).

File: Dockerfile [hunk 15]
```diff
@@ -49,6 +53,32 @@ RUN curl -sL \
...
+    rm -rf /tmp/*
+
+# Install cosign (uses linux-amd64 / linux-arm64 naming)
+RUN curl -sL \
+        -o cosign-linux-${TARGETARCH} \
+        "https://github.com/sigstore/cosign/releases/download/v${COSIGN_VERSION}/cosign-linux-${TARGETARCH}" && \
+    curl -sL \
+        -o cosign_checksums.txt \
+        "https://github.com/sigstore/cosign/releases/download/v${COSIGN_VERSION}/cosign_checksums.txt" && \
+    sha256sum --ignore-missing -c cosign_checksums.txt && \
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply Dockerfile]: Fixed in 7c467b3 — switched to `curl -fsSL` for crane and cosign downloads so HTTP errors cause build failures.

File: tests/test_chainguard.py [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+    def test_crane_not_available(self, mock_which):
+        result = detect_chainguard_image("cgr.dev/chainguard/python:latest")
+        assert result is None
+
+
+class TestDetectFromProvenance:
+    @patch("sbomify_action._generation.chainguard.shutil.which", return_value="/usr/local/bin/crane")
+    @patch("sbomify_action._generation.chainguard._run_crane")
+    def test_detects_chainguard_base_in_provenance(self, mock_crane, mock_which):
+        # Resolve platform digest for the found Chainguard image
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply tests/test_chainguard.py:315]: Added in 7c467b3 — new `test_detects_chainguard_with_registry_port` test that uses `localhost:5000/myorg/myapp:v1.0` and verifies the crane calls use the correct repo reference.

File: sbomify_action/cli/main.py [hunk 17]
```diff
@@ -1127,15 +1129,83 @@ def run_pipeline(config: Config) -> None:
...
+                            "Any binaries added via COPY --from=... (e.g. cosign, crane, osv-scanner) "
+                            "will NOT be included. Use ADDITIONAL_PACKAGES to declare them. "
+                            "See: https://github.com/sbomify/sbomify-action#additional-packages",
+                            title="Chainguard Image Detected",
+                        )
+                        if config.sbom_format == "cyclonedx":
+                            spec = config.spec_version or "1.6"
+                            # Chainguard conversion requires CDX 1.3+; fall back for older versions
+                            if spec < "1.3":
+                                logger.info(
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py]: Fixed in 6aa9357 — now uses tuple comparison (`tuple(int(x) for x in spec.split("."))`) for correct numeric ordering.

File: sbomify_action/cli/main.py [hunk 18]
```diff
@@ -1127,15 +1129,83 @@ def run_pipeline(config: Config) -> None:
...
+                                    f"CycloneDX {spec} requested; Chainguard conversion requires 1.3+ — "
+                                    "falling back to normal generation"
+                                )
+                                chainguard_info = None
+                            else:
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py]: Fixed in 6aa9357 — `gha_warning()` now only emits after all checks pass and the SBOM is written.

File: sbomify_action/cli/main.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py]: Fixed in 6aa9357 — version compatibility is now checked before `fetch_chainguard_sbom()`, avoiding the unnecessary network call.

File: sbomify_action/tool_checks.py [hunk 19]
```diff
@@ -83,6 +83,28 @@ class ToolInfo:
...
+        "name": "cosign",
+        "description": "Container image signing and attestation tool",
+        "install_instructions": (
+            "Install via package manager:\n"
+            "  - macOS: brew install cosign\n"
+            "  - Linux: See https://github.com/sigstore/cosign/releases"
+        ),
+        "homepage": "https://github.com/sigstore/cosign",
+        "required_for": ["Chainguard SBOM retrieval"],
+    },
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/tool_checks.py:107]: By design — `_get_external_tools()` enumerates SBOM generators, while crane/cosign are used for Chainguard detection (an optimization, not a generation path). Their availability is checked at call time in `detect_chainguard_image()`/`fetch_chainguard_sbom()` with graceful fallback. The metadata entries in `_TOOL_METADATA` are for install instructions when users explicitly query tool status.

File: sbomify_action/_generation/chainguard.py [hunk 20]
```diff
@@ -0,0 +1,474 @@
...
+            continue
+
+        payload_b64 = envelope.get("payload", "")
+        if not payload_b64:
+            continue
+
+        try:
+            payload = json.loads(base64.b64decode(payload_b64))
+        except (json.JSONDecodeError, ValueError, UnicodeDecodeError):
+            continue
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/_generation/chainguard.py:345]: Fixed in bc076c5 — added `binascii.Error` to the exception handler.

File: sbomify_action/tool_checks.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/tool_checks.py:107]: Same as the previous comment on this topic — by design. The metadata entries serve as documentation for `get_tool_install_message()` when users explicitly query tool status. crane/cosign aren't generators and don't belong in the generator registry.

File: sbomify_action/cli/main.py [hunk 21]
```diff
@@ -1127,15 +1129,85 @@ def run_pipeline(config: Config) -> None:
...
+                if chainguard_info and config.sbom_format == "cyclonedx":
+                    spec = config.spec_version or "1.6"
+                    # Compare as tuples for correct numeric ordering (e.g. "1.10" > "1.3")
+                    spec_parts = tuple(int(x) for x in spec.split("."))
+                    if spec_parts < (1, 3):
+                        logger.info(
+                            f"CycloneDX {spec} requested; Chainguard conversion requires 1.3+ — "
+                            "falling back to normal generation"
+                        )
+                        chainguard_info = None
```
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/cli/main.py]: Fixed in 48b10b9 — compatibility checks now run before `detect_chainguard_image()`, skipping detection entirely when the format/version is incompatible.

File: sbomify_action/tool_checks.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (vpetersson) [reply sbomify_action/tool_checks.py:107]: Same point as previous rounds — by design. These entries serve `get_tool_install_message()` for explicit queries. crane/cosign are not generators and the architecture intentionally separates detection tools from generation tools.
